hypervisor/arch/x86/guest/vm.c

```diff
@@ -239,15 +239,17 @@ static void prepare_prelaunched_vm_memmap(struct acrn_vm *vm, const struct acrn_
 					|| (entry->type == E820_TYPE_ACPI_NVS)) {
 				ept_add_mr(vm, (uint64_t *)vm->arch_vm.nworld_eptp, base_hpa, entry->baseaddr,
 					entry->length, EPT_RWX | EPT_WB);
+				base_hpa += entry->length;
+				remaining_hpa_size -= entry->length;
 			}
 
 			/* GPAs under 1MB are always backed by physical memory */
 			if ((entry->type != E820_TYPE_RAM) && (entry->baseaddr < (uint64_t)MEM_1M)) {
 				ept_add_mr(vm, (uint64_t *)vm->arch_vm.nworld_eptp, base_hpa, entry->baseaddr,
 					entry->length, EPT_RWX | EPT_UNCACHED);
+				base_hpa += entry->length;
+				remaining_hpa_size -= entry->length;
```

Comment:
The action is same. Seems we can merge the 2 "if () {}". Can we?

Reply:
I tried to merge the two cases, but seems we will have one more LoC:

diff --git a/hypervisor/arch/x86/guest/vm.c b/hypervisor/arch/x86/guest/vm.c
index 423d4335..600bf043 100644
--- a/hypervisor/arch/x86/guest/vm.c
+++ b/hypervisor/arch/x86/guest/vm.c
@@ -228,6 +228,7 @@ static void prepare_prelaunched_vm_memmap(struct acrn_vm *vm, const struct acrn_
 
        for (i = 0U; i < vm->e820_entry_num; i++) {
                const struct e820_entry *entry = &(vm->e820_entries[i]);
+               uint64_t prot = 0UL;
 
                if (entry->length == 0UL) {
                        break;
@@ -237,17 +238,19 @@ static void prepare_prelaunched_vm_memmap(struct acrn_vm *vm, const struct acrn_
                        /* Do EPT mapping for GPAs that are backed by physical memory */
                        if ((entry->type == E820_TYPE_RAM) || (entry->type == E820_TYPE_ACPI_RECLAIM)
                                        || (entry->type == E820_TYPE_ACPI_NVS)) {
-                               ept_add_mr(vm, (uint64_t *)vm->arch_vm.nworld_eptp, base_hpa, entry->baseaddr,
-                                       entry->length, EPT_RWX | EPT_WB);
+                               prot = EPT_RWX | EPT_WB;
                        }
 
                        /* GPAs under 1MB are always backed by physical memory */
                        if ((entry->type != E820_TYPE_RAM) && (entry->baseaddr < (uint64_t)MEM_1M)) {
+                               prot = EPT_RWX | EPT_UNCACHED;
+                       }
+                       if (prot != 0UL) {
                                ept_add_mr(vm, (uint64_t *)vm->arch_vm.nworld_eptp, base_hpa, entry->baseaddr,
-                                       entry->length, EPT_RWX | EPT_UNCACHED);
+                                       entry->length, prot);
+                               base_hpa += entry->length;
+                               remaining_hpa_size -= entry->length;
                        }
-                       base_hpa += entry->length;
-                       remaining_hpa_size -= entry->length;
                } else if (entry->type == E820_TYPE_RAM) {
                        pr_warn("%s: HPA size incorrectly configured in v820\n", __func__);
                }

Reply:
I have tried to merge the code like below, but this would have one more LoC, how do you think?

	for (i = 0U; i < vm->e820_entry_num; i++) {
		const struct e820_entry *entry = &(vm->e820_entries[i]);
		uint64_t prot = 0UL;

		if (entry->length == 0UL) {
			break;
		}

		if (remaining_hpa_size >= entry->length) {
			/* Do EPT mapping for GPAs that are backed by physical memory */
			if ((entry->type == E820_TYPE_RAM) || (entry->type == E820_TYPE_ACPI_RECLAIM)
					|| (entry->type == E820_TYPE_ACPI_NVS)) {
				prot = EPT_RWX | EPT_WB;
			}

			/* GPAs under 1MB are always backed by physical memory */
			if ((entry->type != E820_TYPE_RAM) && (entry->baseaddr < (uint64_t)MEM_1M)) {
				prot = EPT_RWX | EPT_UNCACHED;
			}
			if (prot != 0UL) {
				ept_add_mr(vm, (uint64_t *)vm->arch_vm.nworld_eptp, base_hpa, entry->baseaddr,
					entry->length, prot);
				base_hpa += entry->length;
				remaining_hpa_size -= entry->length;
			}
		} else if (entry->type == E820_TYPE_RAM) {
			pr_warn("%s: HPA size incorrectly configured in v820\n", __func__);
		}